client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx

```diff
@@ -98,15 +99,25 @@ const PatternAssembler: Step = ( { navigation } ) => {
 	};
 
 	const addSection = ( pattern: Pattern ) => {
+		incrementIndexRef.current++;
 		if ( sectionPosition !== null ) {
 			setSections( [
 				...sections.slice( 0, sectionPosition ),
-				pattern,
+				{
+					...pattern,
+					key: sections[ sectionPosition ].key,
+				},
 				...sections.slice( sectionPosition + 1 ),
 			] );
 			setScrollToSelectorByPosition( sectionPosition );
 		} else {
-			setSections( [ ...( sections as Pattern[] ), pattern ] );
+			setSections( [
+				...( sections as Pattern[] ),
+				{
+					...pattern,
+					key: `${ incrementIndexRef.current }-${ pattern.id }`,
```

Comment:
I'm looking at this code after reading the P2 about keys and animations, and have some suggestions:

The code that manages the `sections` array and performs add/delete/move operations on it, it would be a nice use case for `useReducer`. Implement all the operations in the reducer, and then just dispatch actions like `{ type: 'move', position }`.

The auto-increment counter could be part of the reducer state, and incremented only on an "add" action.

Also, it seems that the `pattern.id` is not unique enough? There can be multiple patterns in the list with the same `id`?

The `pattern.id` doesn't need to be part of the key, as the counter value is already unique. The key can be just `incrementIndexRef.current`. The key can be a number, doesn't need to be stringified.

Reply:
> nice use case for useReducer.
Good suggestion! It may needs to take me some time on the implementation
> it seems that the pattern.id is not unique enough? There can be multiple patterns in the list with the same id
Yes it is
> the counter value is already unique
I think you're right about this